packages/nest/src/utils/ast-utils.ts

```diff
@@ -0,0 +1,480 @@
+      if (text.match('^\r?\r?\n')) {
+        toInsert = `,${
+          text.match(/^\r?\n\s+/)[0]
+        }${metadataField}: [${expression}]`;
+      } else {
+        toInsert = `, ${metadataField}: [${expression}]`;
+      }
+    }
+  } else if (!isArray && node.kind == ts.SyntaxKind.ArrayLiteralExpression) {
+    // We found the field but it's empty. Insert it just before the `]`.
+    position--;
+    toInsert = `${expression}`;
+  } else {
+    // Get the indentation of the last element, if any.
+    const text = node.getFullText(source);
+    if (text.match(/^\r?\n/)) {
+      toInsert = `,${text.match(/^\r?\n(\r?)\s+/)[0]}${expression}`;
+    } else {
+      toInsert = `, ${expression}`;
+    }
+  }
+  const insert = new InsertChange(ngModulePath, position, toInsert);
+  return [insert];
+}
+
+export function addImportToModule(
+  source: ts.SourceFile,
+  modulePath: string,
+  symbolName: string
+): Change[] {
+  return _addSymbolToNestModuleMetadata(
+    source,
+    modulePath,
+    'imports',
+    symbolName
+  );
+}
+
+export function addProviderToModule(
+  source: ts.SourceFile,
+  modulePath: string,
+  symbolName: string
+): Change[] {
+  return _addSymbolToNestModuleMetadata(
+    source,
+    modulePath,
+    'providers',
+    symbolName
+  );
+}
+
+function _nestImportsFromNode(
+  node: ts.ImportDeclaration,
+  _sourceFile: ts.SourceFile
+): { [name: string]: string } {
+  const ms = node.moduleSpecifier;
+  let modulePath: string;
+  switch (ms.kind) {
+    case ts.SyntaxKind.StringLiteral:
+      modulePath = (ms as ts.StringLiteral).text;
+      break;
+    default:
+      return {};
+  }
+
+  if (!modulePath.startsWith('@nestjs/')) {
+    return {};
+  }
+
+  if (node.importClause) {
+    if (node.importClause.name) {
+      // This is of the form `import Name from 'path'`. Ignore.
+      return {};
+    } else if (node.importClause.namedBindings) {
+      const nb = node.importClause.namedBindings;
+      if (nb.kind == ts.SyntaxKind.NamespaceImport) {
+        // This is of the form `import * as name from 'path'`. Return `name.`.
+        return {
+          [(nb as ts.NamespaceImport).name.text + '.']: modulePath,
+        };
+      } else {
+        // This is of the form `import {a,b,c} from 'path'`
+        const namedImports = nb as ts.NamedImports;
+
+        return namedImports.elements
+          .map((is: ts.ImportSpecifier) =>
+            is.propertyName ? is.propertyName.text : is.name.text
+          )
+          .reduce((acc: { [name: string]: string }, curr: string) => {
+            acc[curr] = modulePath;
+
+            return acc;
+          }, {});
+      }
+    }
+
+    return {};
+  } else {
+    // This is of the form `import 'path';`. Nothing to do.
+    return {};
+  }
+}
+
+export function getDecoratorMetadata(
+  source: ts.SourceFile,
+  identifier: string,
+  module: string
+): ts.Node[] {
+  const angularImports: { [name: string]: string } = findNodes(
```

Comment:
I agree with what Jason said about these utils, we shouldn't have to copy the whole file from angular (or we get things like this ☝🏼). Can these be put into the workspace package, and then have angular and nest import from there?